File: backend/src/Krakenar.Contracts/Configurations/Configuration.cs
using Krakenar.Contracts.Actors;
using Krakenar.Contracts.Logging;
using Krakenar.Contracts.Settings;

namespace Krakenar.Contracts.Configurations;

public class Configuration : Aggregate
{
  [JsonIgnore]
  public string? Secret { get; set; }
  public Actor SecretChangedBy { get; set; } = new();
  public DateTime SecretChangedOn { get; set; }

  public UniqueNameSettings UniqueNameSettings { get; set; } = new();
  public PasswordSettings PasswordSettings { get; set; } = new();
  public LoggingSettings LoggingSettings { get; set; } = new();
}

File: frontend/src/types/configuration.ts
import type { Actor } from "./actor";
import type { Aggregate } from "./aggregate";
import type { Change } from "./change";
import type { LoggingSettings, PasswordSettings, UniqueNameSettings } from "./settings";

export type Configuration = Aggregate & {
  secretChangedBy: Actor;
  secretChangedOn: string;
  uniqueNameSettings: UniqueNameSettings;
  passwordSettings: PasswordSettings;
  loggingSettings: LoggingSettings;
};

export type ReplaceConfigurationPayload = {
  uniqueNameSettings: UniqueNameSettings;
  passwordSettings: PasswordSettings;
  loggingSettings: LoggingSettings;
};

export type UpdateConfigurationPayload = {
  secret?: Change<string>;
  uniqueNameSettings?: UniqueNameSettings;
  passwordSettings?: PasswordSettings;
  loggingSettings?: LoggingSettings;
};
